addon/components/suspense-component/index.js

```diff
@@ -0,0 +1,108 @@
+import { getOwner } from "@ember/application";
+import Component from "@glimmer/component";
+import Ember from "ember";
+import IS_BROWSER from "ember-async-component/utils/is-browser";
+import { tracked } from "@glimmer/tracking";
+
+/**
+ * This is the suspense component which be used to by container components when making API calls in a component.
+ * This component handles server side rendering issues and loading and error states out of the box for the consumer
+ * @param {Function|object} [promise] Required promise for the component to render the loading, success and error state
+ * @param {boolean} [blockRender] Default is false. Used for deciding if the fastboot server should wait for the API call to complete
+ * ...
+ * @example
+ *   <SuspenseComponent
+ *     @promise={{this.promise}}
+ *     @blockRender={{false}}
+ *     as |task|
+ *   >
+ *     {{#if task.isLoading}}
+ *       Loading...
+ *     {{else if task.isSuccess}}
+ *       {{task.data.userRequest.name}}: {{task.data.userRequest.time}}
+ *     {{else if task.isError}}
+ *       Error occurred: {{task.errorReason}}
+ *     {{/if}}
+ *   </SuspenseComponent>
+ */
+
+class Task {
+  @tracked data = null;
+  @tracked errorReason = null;
+  @tracked isLoading = false;
+
+  get isError() {
+    return Boolean(this.errorReason);
+  }
+
+  get isSuccess() {
+    return Boolean(this.data);
+  }
+}
+
+export default class SuspenseComponent extends Component {
+  blockRender = false;
+  promise = null;
+
+  constructor() {
+    super(...arguments);
+
+    // Recommended way to get the service in the addons and not forcing fastboot for every consuming application
+    // https://ember-fastboot.com/docs/addon-author-guide#accessing-the-fastboot-service
+    this.fastboot = getOwner(this).lookup("service:fastboot");
+  }
+
+  get data() {
+    const blockRender = this.args.blockRender || false;
+    const promise = this.args.promise;
+
+    return this.execute(promise, blockRender);
+  }
+
+  execute(promiseArg, blockRender) {
+    const task = new Task();
+    const promiseOrCallback = promiseArg;
+    let promise;
+
+    if (typeof promiseOrCallback === "function") {
+      promise = promiseOrCallback();
+    } else {
+      promise = promiseOrCallback;
+    }
+
+    if (this.promise === promise) {
+      return this.task;
+    }
+
+    this.task = task;
+    this.promise = promise;
+    if (!IS_BROWSER && !blockRender) {
+      // we are not supposed to block rendering on the server
+      return null;
+    }
+
+    task.isLoading = true;
+
+    if (blockRender && this.fastboot.isFastBoot) {
```

Comment:
I think this would fail hard if `ember-cli-fastboot` was **not** installed, right? If `ember-cli-fastboot` is not present and `@blockRender={{true}}` were passed this would error because `this.fastboot` would be undefined?

Reply:
It should probably fail hard or at least fail earlier than this line. At the end blockRender has specific meaning and allowing the parameter to be specified without fastboot may confuse people.

Reply:
Seems fine to assert, but it should do that **much** earlier (so that the stack trace is somewhat useful).